File: scripts/transport-v4-crypto-usage.mjs
import assert from "node:assert/strict";
import { types } from "node:util";
import { VectorError } from "./transport-v4-codec.mjs";

const requireThat = (condition, code) => { if (!condition) throw new VectorError(code); };

export function verifyCryptoUsageRegistry(schema) {
  const registry = schema.crypto_usage_registry;
  assert.equal(registry.status, "design_fixed_limits_reference_charges_only");
  assert.equal(registry.source, "3.5.3");
  assert.equal(registry.quantity_max, "18446744073709551615");
  assert.equal(registry.authentication_block_bytes, 16);
  assert.equal(registry.length_blocks, 1);
  assert.deepEqual(registry.charge_fields, ["operation", "aad_bytes", "input_bytes"]);
  assert.deepEqual(registry.operations, ["seal", "open"]);
  assert.deepEqual(Object.keys(registry.profiles).sort(), Object.keys(schema.crypto_profiles).sort());
  // Assert the design's fixed table, not a deployment-selectable budget.
  const count = power => (1n << BigInt(power)).toString();
  const limits = (calls, blocks, bytes) => ({ seal_calls: count(calls), open_attempts: count(calls), authentication_blocks: count(blocks), ciphertext_bytes: count(bytes) });
  for (const [profile, usage] of Object.entries(registry.profiles)) {
    const algorithm = schema.crypto_profiles[profile].dh_algorithm;
    assert.ok(algorithm === 0 || algorithm === 1);
    assert.deepEqual(usage, {
      key: limits(20, algorithm === 0 ? 32 : 31, 36),
      epoch: limits(algorithm === 0 ? 31 : 30, algorithm === 0 ? 42 : 41, algorithm === 0 ? 46 : 45),
      session: limits(algorithm === 0 ? 36 : 35, algorithm === 0 ? 52 : 51, algorithm === 0 ? 56 : 55),
      root_max_age_ms: "86400000", max_epochs: count(16), max_record_key_derivations: count(28),
    });
    assert.equal(schema.crypto_profiles[profile].tag_bytes, 16);
  }
  assert.match(registry.qualification, /no independent aggregate cryptographic proof/u);
}

// Decimal-string charge arithmetic for public fixtures. This performs no AEAD,
// checks no actual frame/profile input limit and reserves no crypto usage. The
// original owner must apply every key/epoch/Session cap atomically before work,
// never refund a precharge, and revalidate root age at actual invocation.
export function recordUsageReference(schema, profile, input) {
  verifyCryptoUsageRegistry(schema);
  const registry = schema.crypto_usage_registry;
  requireThat(typeof profile === "string" && Object.hasOwn(registry.profiles, profile), "crypto_usage_profile");
  requireThat(input !== null && typeof input === "object" && !types.isProxy(input) &&
    [Object.prototype, null].includes(Object.getPrototypeOf(input)), "crypto_usage_input");
  const keys = Reflect.ownKeys(input), names = registry.charge_fields;
  requireThat(keys.length === names.length && keys.every(key => names.includes(key)), "crypto_usage_fields");
  const captured = Object.create(null);
  for (const key of names) {
    const descriptor = Object.getOwnPropertyDescriptor(input, key);
    requireThat(descriptor && Object.hasOwn(descriptor, "value") && descriptor.enumerable, "crypto_usage_data");
    captured[key] = descriptor.value;
  }
  requireThat(typeof captured.operation === "string" && registry.operations.includes(captured.operation), "crypto_usage_operation");
  const maximum = BigInt(registry.quantity_max);
  const integer = value => {
    requireThat(typeof value === "string" && value.length <= registry.quantity_max.length && /^(?:0|[1-9][0-9]*)$/u.test(value), "crypto_usage_integer");
    const n = BigInt(value); requireThat(n <= maximum, "crypto_usage_integer"); return n;
  };
  const checked = n => { requireThat(n >= 0n && n <= maximum, "crypto_usage_overflow"); return n; };
  const aad = integer(captured.aad_bytes), bytes = integer(captured.input_bytes);
  const tag = BigInt(schema.crypto_profiles[profile].tag_bytes), block = BigInt(registry.authentication_block_bytes);
  const seal = captured.operation === "seal";
  if (!seal) requireThat(bytes >= tag, "crypto_usage_ciphertext");
  const payload = seal ? bytes : bytes - tag, ciphertext = seal ? checked(bytes + tag) : bytes;
  const blocks = n => n / block + (n % block === 0n ? 0n : 1n);
  const authentication = checked(checked(blocks(aad) + blocks(payload)) + BigInt(registry.length_blocks));
  return Object.freeze({ calls: "1", authentication_blocks: authentication.toString(), ciphertext_bytes: ciphertext.toString() });
}

export function buildCryptoUsageCorpus(schema) {
  verifyCryptoUsageRegistry(schema);
  const ids = new Set(), vectors = [];
  for (const profile of Object.keys(schema.crypto_usage_registry.profiles)) {
    for (const spec of schema.crypto_usage_vector_plan) {
      assert.match(spec.id, /^[a-z][a-z0-9_]*$/u);
      const id = `crypto_usage_${schema.crypto_profiles[profile].dh_algorithm}_${spec.id}`;
      assert.ok(!ids.has(id)); ids.add(id);
      const input = structuredClone(spec.input);
      if (spec.expected_error) {
        assert.equal(spec.expected, undefined);
        assert.throws(() => recordUsageReference(schema, profile, input), error => error instanceof VectorError && error.code === spec.expected_error, id);
        vectors.push({ id, profile, input, expected_error: spec.expected_error });
      } else {
        const expected = recordUsageReference(schema, profile, input);
        assert.deepEqual(expected, spec.expected, id);
        vectors.push({ id, profile, input, expected });
      }
    }
  }
  return { status: "draft", schema_revision: schema.schema_revision, design_sha256: schema.design_sha256,
    coverage: "fixed_crypto_usage_limits_and_reference_record_charges_only", qualification: schema.crypto_usage_registry.qualification, vectors };
}

File: flowersec-rust/tests/transport_v4_crypto_usage.rs
//! Independent test-only charges, not input eligibility or usage reservation.
#[allow(dead_code)]
#[path = "../src/protocol_v4_registry_generated.rs"]
mod registry;
use serde_json::{Value, json};

fn charge(
    spec: &Value,
    profiles: &Value,
    profile: &str,
    input: &Value,
) -> Result<Value, &'static str> {
    if spec["profiles"].get(profile).is_none() {
        return Err("crypto_usage_profile");
    }
    let fields = spec["charge_fields"].as_array().unwrap();
    let object = input.as_object().ok_or("crypto_usage_input")?;
    if object.len() != fields.len()
        || object
            .keys()
            .any(|key| !fields.iter().any(|f| f.as_str() == Some(key)))
    {
        return Err("crypto_usage_fields");
    }
    let operation = input["operation"]
        .as_str()
        .ok_or("crypto_usage_operation")?;
    if !spec["operations"]
        .as_array()
        .unwrap()
        .iter()
        .any(|v| v.as_str() == Some(operation))
    {
        return Err("crypto_usage_operation");
    }
    let maximum = spec["quantity_max"]
        .as_str()
        .unwrap()
        .parse::<u64>()
        .unwrap();
    let integer = |value: &Value| -> Result<u64, &'static str> {
        let s = value.as_str().ok_or("crypto_usage_integer")?;
        if s.is_empty()
            || s.len() > spec["quantity_max"].as_str().unwrap().len()
            || s.len() > 1 && s.starts_with('0')
            || !s.bytes().all(|c| c.is_ascii_digit())
        {
            return Err("crypto_usage_integer");
        }
        let n = s.parse::<u64>().map_err(|_| "crypto_usage_integer")?;
        if n > maximum {
            return Err("crypto_usage_integer");
        }
        Ok(n)
    };
    let aad = integer(&input["aad_bytes"])?;
    let bytes = integer(&input["input_bytes"])?;
    let tag = profiles[profile]["tag_bytes"].as_u64().unwrap();
    let (payload, ciphertext) = if operation == "seal" {
        (
            bytes,
            bytes.checked_add(tag).ok_or("crypto_usage_overflow")?,
        )
    } else {
        (
            bytes.checked_sub(tag).ok_or("crypto_usage_ciphertext")?,
            bytes,
        )
    };
    let block = spec["authentication_block_bytes"].as_u64().unwrap();
    let blocks = aad
        .div_ceil(block)
        .checked_add(payload.div_ceil(block))
        .and_then(|n| n.checked_add(spec["length_blocks"].as_u64().unwrap()))
        .ok_or("crypto_usage_overflow")?;
    if ciphertext > maximum || blocks > maximum {
        return Err("crypto_usage_overflow");
    }
    Ok(
        json!({"calls":"1","authentication_blocks":blocks.to_string(),"ciphertext_bytes":ciphertext.to_string()}),
    )
}
fn spec() -> (Value, Value) {
    (
        serde_json::from_str(registry::CRYPTO_USAGE_REGISTRY_JSON).unwrap(),
        serde_json::from_str(registry::CRYPTO_PROFILES_JSON).unwrap(),
    )
}
#[test]
fn all_shared_usage_vectors() {
    let (spec, profiles) = spec();
    let corpus: Value = serde_json::from_str(include_str!(
        "../../testdata/transport_v4/crypto_usage.json"
    ))
    .unwrap();
    let vectors = corpus["vectors"].as_array().unwrap();
    assert_eq!(vectors.len(), 28);
    for v in vectors {
        let original = v["input"].clone();
        let actual = charge(&spec, &profiles, v["profile"].as_str().unwrap(), &original);
        if let Some(error) = v["expected_error"].as_str() {
            assert_eq!(actual, Err(error), "{}", v["id"]);
        } else {
            assert_eq!(actual, Ok(v["expected"].clone()), "{}", v["id"]);
        }
        assert_eq!(original, v["input"]);
    }
}
#[test]
fn block_boundaries_and_strict_quantities() {
    let (spec, profiles) = spec();
    for profile in spec["profiles"].as_object().unwrap().keys() {
        for aad in 0u64..=33 {
            for payload in 0u64..=33 {
                let seal=charge(&spec,&profiles,profile,&json!({"operation":"seal","aad_bytes":aad.to_string(),"input_bytes":payload.to_string()})).unwrap();
                let open=charge(&spec,&profiles,profile,&json!({"operation":"open","aad_bytes":aad.to_string(),"input_bytes":(payload+16).to_string()})).unwrap();
                assert_eq!(seal, open);
                assert_eq!(
                    seal["authentication_blocks"],
                    json!((aad.div_ceil(16) + payload.div_ceil(16) + 1).to_string())
                );
            }
        }
        for bad in [
            json!(1),
            json!(true),
            Value::Null,
            json!(""),
            json!("01"),
            json!("-1"),
            json!("+1"),
            json!("1.0"),
            json!("1\n"),
            json!("١"),
            json!("18446744073709551616"),
        ] {
            assert_eq!(
                charge(
                    &spec,
                    &profiles,
                    profile,
                    &json!({"operation":"seal","aad_bytes":bad,"input_bytes":"0"})
                ),
                Err("crypto_usage_integer")
            );
        }
    }
}
